src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs

```diff
@@ -111,6 +111,39 @@ public static ITypeReference GetEnumType(this ITypeDefinition type)
             throw new InvalidOperationException("All enums should have a value__ field!");
         }
 
+        public static bool IsOrContainsReferenceType(this ITypeReference type)
+        {
+            Queue<ITypeReference> typesToCheck = new Queue<ITypeReference>();
+            HashSet<ITypeReference> visited = new HashSet<ITypeReference>();
+
+            typesToCheck.Enqueue(type);
+
+            while (typesToCheck.Count != 0)
+            {
+                var typeToCheck = typesToCheck.Dequeue();
+                visited.Add(typeToCheck);
+
+                if (!typeToCheck.IsValueType)
+                    return true;
+
+                var resolvedType = typeToCheck.ResolvedType;
+
+                // If it is dummy we cannot really check so assume it does because that is will be the most conservative 
+                if (resolvedType is Dummy)
+                    return true;
+
+                foreach (var field in resolvedType.Fields)
```

Comment:
Just to be sure - are these all fiends or only instance fields? (we should only care about instance fields)

Reply:
That is all fields. I can filter out the statics.

Reply:
I think we should filter out statics. Presence of object static fields does not make a struct managed.

---
In reply to: [160838104](https://github.com/dotnet/buildtools/pull/1859#discussion_r160838104) [](ancestors = 160838104)